src/runtime-rs/crates/agent/src/vsock.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::os::unix::prelude::AsRawFd;
+
+use anyhow::{anyhow, Context, Result};
+use tokio::{
+    io::{AsyncBufReadExt, AsyncWriteExt, BufReader},
+    net::UnixStream,
+};
+
+#[derive(Clone, Debug)]
+pub(crate) struct VsockConnectConfig {
+    uds: String,
+    port: u32,
+    dial_timeout_ms: u64,
+    reconnect_timeout_ms: u64,
+}
+
+impl VsockConnectConfig {
+    pub(crate) fn new(
+        uds: &str,
+        port: u32,
+        dial_timeout_ms: u64,
+        reconnect_timeout_ms: u64,
+    ) -> Self {
+        Self {
+            uds: uds.to_string(),
+            port,
+            dial_timeout_ms,
+            reconnect_timeout_ms,
+        }
+    }
+}
+
+pub(crate) async fn connect_vsock(config: &VsockConnectConfig) -> Result<UnixStream> {
+    let retry_times = config.reconnect_timeout_ms / config.dial_timeout_ms;
+    for i in 0..retry_times {
+        match connect_helper(config).await {
+            Ok(stream) => {
+                info!(
+                    sl!(),
+                    "connect success on {} current client fd {}",
+                    i,
+                    stream.as_raw_fd()
+                );
+                return Ok(stream);
+            }
+            Err(err) => {
+                debug!(sl!(), "connect on {} err : {:?}", i, err);
+                tokio::time::sleep(std::time::Duration::from_millis(config.dial_timeout_ms)).await;
+                continue;
+            }
+        }
+    }
+    Err(anyhow!("cannot connect to agent ttrpc server"))
+}
+
+async fn connect_helper(config: &VsockConnectConfig) -> Result<UnixStream> {
+    info!(sl!(), "connect vsock {:?}", &config);
+    let mut stream = UnixStream::connect(&config.uds).await.context("connect")?;
```

Comment:
So it only supports hybrid vsock that clh/firecracker has? For qemu, we still need to connect to vhost vsock that is of format <cid:port>.

Reply:
Thanks for your review,  update to support both `vsock` and `hvsock`.